### ISSUE TITLE
VSTest does not run tests in X64 when specified in Platform.

### ISSUE THREAD

**Issue by andradf** [Microsoft/vso-agent-tasks#1252]
I get the following error in VSO when trying to run tests in an x64 assembly.

```
Test run will use DLL(s) built for framework Framework45 and platform X86. Following DLL(s) will not be part of run: 
 
My.Tests.dll, My.Tests.dll are built for Framework Framework45 and Platform X64.
 
##[error]Error: None of the provided test containers match the Platform Architecture and .Net Framework settings for the test run. Platform: X86  .Net Framework: Framework45. Go to http://go.microsoft.com/fwlink/?LinkID=330428 for more details on managing these settings.
 
##[error]
 
##[error]VSTest Test Run failed with exit code: 1

```
In VSTest.ps1 when Invoke-VsTest is called, the variable $platform is not even passed to the cmdlet.

`Invoke-VSTest -TestAssemblies $testAssemblyFiles -VSTestVersion $vsTestVersion -TestFiltercriteria $testFiltercriteria -RunSettingsFile $runSettingsFileWithParallel -PathtoCustomTestAdapters $pathtoCustomTestAdapters -CodeCoverageEnabled $codeCoverage -OverrideTestrunParameters $overrideTestrunParameters -OtherConsoleOptions $otherConsoleOptions -WorkingFolder $workingDirectory -TestResultsFolder $testResultsDirectory -SourcesDirectory $sourcesDirectory`

**Comment by atinb** [comment]
The Platform field in the task is purely for reporting purposes. You need to mention /platform:x64 in additional command line options or add this setting in your runsettings file using TargetPlatform tag

**Comment by aligneddev** [comment]
I'm getting this error from VS 2017 v15.4.3.

From [the runsettings documentation example ](https://msdn.microsoft.com/en-us/library/jj635153.aspx)
"you can also change it from menu **Test, Test Settings, Default Processor Architecture**".

Maybe that will help someone.

**Comment by jemiller0** [comment]
I just upgraded to VS 2017 15.5 and now I'm getting the error. It was working fine before the upgrade. Whatever "improvements" were made to testing in 15.5 seem to have broken it for me.

**Comment by aphilpotts** [comment]
I am seeing something like this too, here full error message:

Test run will use DLL(s) built for framework .NETFramework,Version=v4.5 and platform X64. Following DLL(s) do not match framework/platform settings.
LI.WebServer.UnitTest.dll is built for Framework 4.6.2 and Platform X64.

The command line I use is:

"C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\IDE\CommonExtensions\Microsoft\TestWindow\vstest.console.exe"  "Logical Ink\Server\LI.WebServer.UnitTest\bin\x64\debug\LI.WebServer.UnitTest.dll" /Settings:"Logical Ink\Server\server.runsettings" /Framework:Framework45 /TestAdapterPath:"Logical Ink\Server\LI.WebServer.UnitTest\bin\x64\debug" /TestCaseFilter:"TestCategory^!=Slow" /InIsolation /Platform:X64 /Logger:trx /diag:ServerTestDiagnostics.log

**Comment by aphilpotts** [comment]
Tests run OK from within Visual Studio...

**Comment by jemiller0** [comment]
I don't know if this is related or not, but, a problem that I ran into was that it wasn't detecting the tests. I was able to solve the problem by removing the binding redirects from the App.config files and then regenerating them using `Get-Project -All | Add-BindingRedirect`. This has happened to me in a couple different solutions now.

**Comment by aphilpotts** [comment]
No joy with that, although I didn't do exactly as you said, I did not want to splat all my project files so I only changed the one project with all my tests, no improvement.

FWIW, it fails in Visual Studio too

**Comment by jemiller0** [comment]
You may need to do it to all your projects. Or at least the project that the test project is testing. I deleted the entire `<runtime>` in mine. Then, ran the NuGet command from above and it regenerated the element. If you're using Git, just make sure you have everything committed first. So, you can revert the changes if necessary. I have no idea why the test framework would be so sensitive to binding redirects. To do only one project, you can use something like `Get-Project -Name MyLibrary | Add-BindingRedirect`.

**Comment by southernsun** [comment]
not sure why this issue is closed. Why are experiencing similar behavior and are not able to fix this. We are using VSTS and everything seemed to be working fine until a week ago. We discovered five days after it was broken VSTS was reporting false positive as all tests were "successful" as it didn't find any tests to execute anymore. There are different discussion online about similar issues that don't provide any meaningful solution to this problem.

Our problems started after upgrading the C++ SDK as we updated one of our targets on the C++ project. After getting this error on our build machine we upgraded our Windows SDK:

> Error MSB8036: The Windows SDK version 10.0.16299.0 was not found. Install the required version of Windows SDK or change the SDK version in the project property pages or by right-clicking the solution and selecting "Retarget solution".

Since then the unit tests are no longer discovered. We are running VS 15.6.7 on one machine and 15.7.1 (currently the latest version) on another. We also used the VSTEST Platform Installer in VSTS to get the latest build and use that for the run, but also with no avail.

Our current error in VSTS:
```
2018-05-11T09:13:25.0772361Z ##[section]Starting: VsTest - testAssemblies
2018-05-11T09:13:25.0772361Z ==============================================================================
2018-05-11T09:13:25.0772361Z Task         : Visual Studio Test
2018-05-11T09:13:25.0772361Z Description  : Run tests with Visual Studio test runner
2018-05-11T09:13:25.0772361Z Version      : 1.0.86
2018-05-11T09:13:25.0772361Z Author       : Microsoft Corporation
2018-05-11T09:13:25.0772361Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=624539)
2018-05-11T09:13:25.0772361Z ==============================================================================
2018-05-11T09:13:25.0772361Z Preparing task execution handler.
2018-05-11T09:13:25.6856400Z Executing the powershell script: D:\agent\_work\_tasks\VSTest_ef087383-ee5e-42c7-9a53-ab56c98420f9\1.0.86\VSTest.ps1
2018-05-11T09:13:32.9396865Z D:\agent\_work\_tool\VsTest\15.8.0-preview-20180510-03\x64\tools\net451\Common7\IDE\Extensions\TestPlatform
2018-05-11T09:13:33.7040914Z Working folder: D:\agent\_work\25\s
2018-05-11T09:13:33.7040914Z Executing D:\agent\_work\_tool\VsTest\15.8.0-preview-20180510-03\x64\tools\net451\Common7\IDE\Extensions\TestPlatform\vstest.console.exe "D:\agent\_work\25\s\Modules\Settings\Tests\SettingsTest\bin\x64\Release\FAA.AEE.AEDT.SettingsTest.dll"  /Platform:x64 /Framework:Framework45 /Diag:D:\agent\_work\25\s\verbose_20180511.4_settingsproject.log /logger:trx /TestAdapterPath:"D:\agent\_work\25\s"
2018-05-11T09:13:33.8756925Z Microsoft (R) Test Execution Command Line Tool Version 15.8.0-preview-20180510-03
2018-05-11T09:13:33.8756925Z Copyright (c) Microsoft Corporation.  All rights reserved.
2018-05-11T09:13:33.8756925Z 
2018-05-11T09:13:34.2500949Z Starting test execution, please wait...
2018-05-11T09:13:34.2500949Z Logging Vstest Diagnostics in file: D:\agent\_work\25\s\verbose_20180511.4_settingsproject.log
2018-05-11T09:13:34.3748957Z Test run will use DLL(s) built for framework .NETFramework,Version=v4.5 and platform X64. Following DLL(s) do not match framework/platform settings.
2018-05-11T09:13:34.3748957Z FAA.AEE.AEDT.SettingsTest.dll is built for Framework 4.6.1 and Platform X64.
2018-05-11T09:13:34.3748957Z Go to http://go.microsoft.com/fwlink/?LinkID=236877&clcid=0x409 for more details on managing these settings.
2018-05-11T09:13:34.3748957Z 
2018-05-11T09:13:36.3093081Z Logging TestHost Diagnostics in file: D:\agent\_work\25\s\verbose_20180511.4_settingsproject.host.18-05-11_09-13-35_90370_1.log
2018-05-11T09:13:36.6993106Z No test is available in D:\agent\_work\25\s\Modules\Settings\Tests\SettingsTest\bin\x64\Release\FAA.AEE.AEDT.SettingsTest.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
2018-05-11T09:13:36.7929112Z 
2018-05-11T09:13:36.8709117Z 
2018-05-11T09:13:36.8709117Z 
2018-05-11T09:13:36.8709117Z ##[warning]No results found to publish.
2018-05-11T09:13:36.8865118Z ##[section]Finishing: VsTest - testAssemblies
```

Attached is the associated detailed log from vstest.
[verbose_20180511.4_settingsproject.host.18-05-11_09-13-35_90370_1.log](https://github.com/Microsoft/vsts-tasks/files/1994752/verbose_20180511.4_settingsproject.host.18-05-11_09-13-35_90370_1.log)

I think the main issue is this part:

> TpTrace Information: 0 : 10460, 8, 2018/05/11, 09:13:36.309, 67406343681577, testhost.exe, Execution started.
> TpTrace Verbose: 0 : 10460, 8, 2018/05/11, 09:13:36.324, 67406343830050, testhost.exe, TcpClientExtensions.MessageLoopAsync: Polling on remoteEndPoint: 127.0.0.1:50287 localEndPoint: 127.0.0.1:50288
> TpTrace Verbose: 0 : 10460, 3, 2018/05/11, 09:13:36.324, 67406343839676, testhost.exe, TestPluginCache: Discovering the extensions using extension path.
> TpTrace Information: 0 : 10460, 3, 2018/05/11, 09:13:36.324, 67406343930105, testhost.exe, AssemblyResolver: Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter: Resolving assembly.
> TpTrace Information: 0 : 10460, 3, 2018/05/11, 09:13:36.340, 67406344006230, testhost.exe, AssemblyResolver: Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter: Resolved assembly. 
> TpTrace Information: 0 : 10460, 3, 2018/05/11, 09:13:36.387, 67406344535557, testhost.exe, AssemblyResolver: Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a: Resolving assembly.
> TpTrace Information: 0 : 10460, 3, 2018/05/11, 09:13:36.387, 67406344538212, testhost.exe, AssemblyResolver: Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a: Failed to load assembly.
> TpTrace Verbose: 0 : 10460, 3, 2018/05/11, 09:13:36.387, 67406344547903, testhost.exe, CurrentDomain_AssemblyResolve: Resolving assembly 'Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
> TpTrace Information: 0 : 10460, 3, 2018/05/11, 09:13:36.387, 67406344556253, testhost.exe, AssemblyResolver: Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a: Resolving assembly.
> TpTrace Information: 0 : 10460, 3, 2018/05/11, 09:13:36.387, 67406344556770, testhost.exe, AssemblyResolver: Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a: Resolved from cache.
> TpTrace Verbose: 0 : 10460, 3, 2018/05/11, 09:13:36.387, 67406344557190, testhost.exe, CurrentDomain_AssemblyResolve: Resolving assembly 'Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
> TpTrace Verbose: 0 : 10460, 3, 2018/05/11, 09:13:36.387, 67406344557442, testhost.exe, CurrentDomainAssemblyResolve: Failed to resolve assembly 'Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
> TpTrace Verbose: 0 : 10460, 3, 2018/05/11, 09:13:36.402, 67406344611552, testhost.exe, CurrentDomainAssemblyResolve: Failed to resolve assembly 'Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
> TpTrace Warning: 0 : 10460, 3, 2018/05/11, 09:13:36.418, 67406344776649, testhost.exe, TestPluginDiscoverer: Failed to get types from assembly 'Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.  Skipping test extension scan for this assembly.  Error: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
>    at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
>    at System.Reflection.Assembly.GetTypes()
>    at Microsoft.VisualStudio.TestPlatform.Common.ExtensionFramework.TestPluginDiscoverer.GetTestExtensionsFromAssembly[TPluginInfo,TExtension](Assembly assembly, Dictionary`2 pluginInfos)
> TpTrace Warning: 0 : 10460, 3, 2018/05/11, 09:13:36.418, 67406344799337, testhost.exe, LoaderExceptions: System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
> File name: 'Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' ---> System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
> File name: 'Microsoft.VisualStudio.TestPlatform.TestFramework, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
>    at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
>    at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
>    at System.Reflection.Assembly.Load(AssemblyName assemblyRef)
>    at Microsoft.VisualStudio.TestPlatform.Common.ExtensionFramework.TestPluginCache.CurrentDomainAssemblyResolve(Object sender, AssemblyResolveEventArgs args)
>    at System.AppDomain.OnAssemblyResolveEvent(RuntimeAssembly assembly, String assemblyFullName)
> 
> WRN: Assembly binding logging is turned OFF.
> To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
> Note: There is some performance penalty associated with assembly bind failure logging.
> To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].
> 


Any help on this would be really appreciated.

**Comment by southernsun** [comment]
We opened a new issue to make sure this would get addressed => #7177 

Thank you for your help.
